### PR TITLE
Todoist - MVP widget

### PR DESCRIPTION
**Pull request by faboo03**
WIP - to create the MVP widget for todoist. 

- [x] - added the widget in the menu 
- [x] - Ask for the token for Todoist first
- [ ] - Implement oAuth2 for Todoist and load the official librairy
- [ ] - Get all tasks 
- [ ] - Add stats for query 
- [ ] - Add list for query

Implements: #7

### CODE REVIEW

File: app/package.json [hunk 1]
```diff
@@ -38,7 +38,8 @@
     "redux-persist": "6.0.0",
     "redux-saga": "1.1.3",
     "redux-sagas-injector": "1.1.1",
-    "selectorator": "4.0.3"
+    "selectorator": "4.0.3",
+    "todoist-rest-api": "^1.3.4"
```
Comment (darekkay) [inline app/package.json:42]: Please use exact package versions to make sure that all instances (development, production, CI) use exactly the same versions:

```suggestion
    "todoist-rest-api": "1.3.4"
```

With yarn, you can enforce this with `yarn add package-name --exact` (or `-E`), with npm it's `npm install package-name --save --save-exact`.

File: app/src/components/drawer/index.tsx [hunk 2]
```diff
@@ -36,7 +36,9 @@ const Drawer: React.FC<Props> = ({ addWidgetToLayout }) => {
                 </Button>
               </div>
             )
-          )}
+          ) : 
+            <div className="flex justify-between py-2 italic">{t("widget.common.noWidget")}</div>
```
Comment (darekkay) [inline app/src/components/drawer/index.tsx:40]: I think we shouldn't use a category that has no widgets (that's why the unused are ones commented out). This way, this change (and the according label) are unnecessary. That said, the widget drawer requires [a major redesign](https://dashboard.darekkay.com/docs/roadmap/#features-and-improvements) anyway.

File: app/src/widgets/todoist/properties.ts [hunk 3]
```diff
@@ -0,0 +1,12 @@
+import WidgetCategory from "../categories";
+
+export const widgetType = "todoist";
+export const category = WidgetCategory.Productivity;
+export const initialHeight = 2;
+export const initialWidth = 3;
+export const initialOptions = {
+  token: "xxxxx"
```
Comment (darekkay) [inline app/src/widgets/todoist/properties.ts:8]: The `token` should be empty or undefined by default. That way, you can use `<WidgetUnconfigured>` as long as the token is missing (see the usage in other widgets).

File: app/src/widgets/list.ts [hunk 4]
```diff
@@ -75,6 +75,15 @@ export default {
...
     initialMeta: {}
   },
+  todoist: {
+    configurable: true,
+    widgetType: "todoist",
+    category: "productivity",
+    initialHeight: 3,
+    initialWidth: 4,
+    initialOptions: {},
+    initialMeta: { updateCycle: { hours: 24 } }
```
Comment (darekkay) [inline app/src/widgets/list.ts:85]: This file is autogenerated. Please run `yarn scan-widgets` after any change to `properties.ts`. Currently the values are ouf of sync.
I've looked into ways of improving this workflow by using babel macros, but without ejecting from create-react-app (or forking it), it doesn't get much better.

File: app/src/widgets/todoist/__tests__/index.test.tsx [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+      <Todoist
+        {...widgetProps}
+        id="todoist-mock-id"
+        token="xxxx"
+      />
+    );
+  });
+
+  it("renders without error", () => {
+    expect(wrapper.isEmptyRender()).toBe(false);
```
Comment (darekkay) [inline app/src/widgets/todoist/__tests__/index.test.tsx:22]: FYI - the default (= generated) tests are currently very basic, because I'm planning to replace `enzyme` with `react-testing-library` in the future. If you're experienced with `react-testing-library` (which I'm not that much, yet 😅  ), feel free to write some basic tests. Otherwise you can ignore it for now 😉

File: app/src/widgets/todoist/sagas.ts [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+    yield put(
+      setData({
+        id,
+        values: data
+      })
+    );
+    yield put(updateSuccess(id));
+  } catch (error) {
+    log.error(error);
+    yield put(updateError(id));
```
Comment (darekkay) [inline app/src/widgets/todoist/sagas.ts:36]: Please merge/rebase latest `master` into your branch - I've changed this yesterday a little bit to improve the client-side error handling. After merging/rebasing, this change is required:

```suggestion
    yield put(updateError({ id, error }));
```

File: app/package.json (hunk 1 above)
Comment (darekkay) [inline app/package.json:42]: A second, more generic comment: I've started the project [without any server at all](https://dashboard.darekkay.com/docs/blog/2020-04-11.html#server-module), thinking I could go as far as possible with that approach. However, I've now implemented the first 3rd-party-data-driven widgets (Cryptocurrencies, GitHub Stats). Now I think we should handle all 3rd-party data via the server module for the following reasons:

1. [Request caching](https://dashboard.darekkay.com/docs/blog/2020-04-22.html#request-caching). If the 3rd-party request is implemented on the client, the Todoist API will be hit directly every time the user refreshes the page (page refresh overrides the widget update cycle).

2. When we switch to OAuth later, a developer key will be required. If the request doesn't go through the server module, the developer key will be exposed to the users, which it shouldn't.

In other words, I would move this module from `app/package.json` to `server/package.json` and create internal `/todoist` route(s) as a proxy/facade.

The server module is only a few weeks old, but there are already 3 different routes implemented (2 of them actually used) that you can check as an example.

**Review by darekkay — COMMENTED** [review]
Thanks - the WIP looks good so far 🙂 I've left some comments for you 😉
